phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TaskRegionObserver.java

```diff
@@ -0,0 +1,261 @@
+import org.apache.phoenix.schema.PTable;
+import org.apache.phoenix.schema.PTable.TaskType;
+
+import org.apache.phoenix.util.PhoenixRuntime;
+import org.apache.phoenix.util.QueryUtil;
+
+
+/**
+ * Coprocessor for task related operations. This coprocessor would only be registered
+ * to SYSTEM.TASK table
+ */
+@SuppressWarnings("deprecation")
+public class TaskRegionObserver implements RegionObserver, RegionCoprocessor {
+    public static final Log LOG = LogFactory.getLog(TaskRegionObserver.class);
+    protected ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(TaskType.values().length);
+    private long timeInterval = QueryServicesOptions.DEFAULT_TASK_HANDLING_INTERVAL;
+    private long timeMaxInterval = QueryServicesOptions.DEFAULT_TASK_HANDLING_MAX_INTERVAL;
+    @GuardedBy("TaskRegionObserver.class")
+    // Added for test purposes
+    private long initialDelay = 0;
+
+    @Override
+    public void preClose(final ObserverContext<RegionCoprocessorEnvironment> c,
+            boolean abortRequested) {
+        executor.shutdownNow();
+    }
+
+    @Override
+    public Optional<RegionObserver> getRegionObserver() {
+        return Optional.of(this);
+    }
+
+    @Override
+    public void start(CoprocessorEnvironment env) throws IOException {
+        Configuration config = env.getConfiguration();
+        timeInterval =
+                config.getLong(
+                    QueryServices.TASK_HANDLING_INTERVAL_ATTRIB,
+                    QueryServicesOptions.DEFAULT_TASK_HANDLING_INTERVAL);
+        timeMaxInterval =
+                config.getLong(
+                        QueryServices.TASK_HANDLING_MAX_INTERVAL_ATTRIB,
+                        QueryServicesOptions.DEFAULT_TASK_HANDLING_MAX_INTERVAL);
+        initialDelay =
+                config.getLong(
+                    QueryServices.TASK_HANDLING_INITIAL_DELAY_ATTRIB,
+                    QueryServicesOptions.DEFAULT_TASK_HANDLING_INITIAL_DELAY);
+    }
+    
+    @Override
+    public void postOpen(ObserverContext<RegionCoprocessorEnvironment> e) {
+        final RegionCoprocessorEnvironment env = e.getEnvironment();
+
+        // turn off verbose deprecation logging
+        Logger deprecationLogger = Logger.getLogger("org.apache.hadoop.conf.Configuration.deprecation");
+        if (deprecationLogger != null) {
+            deprecationLogger.setLevel(Level.WARN);
+        }
+        try {
+            Class.forName(PhoenixDriver.class.getName());
+            DropChildViewsTask task = new DropChildViewsTask(e.getEnvironment(), timeMaxInterval);
+            executor.scheduleWithFixedDelay(task, initialDelay, timeInterval, TimeUnit.MILLISECONDS);
+        } catch (ClassNotFoundException ex) {
+            LOG.error("ScheduleTask cannot start!", ex);
+        }
+    }
+
+    public static void addTask(PhoenixConnection conn, TaskType taskType, String tenantId, String schemaName, String tableName)
+            throws IOException {
+        try {
```

Comment:
@twdsilva Do we now require WRITE access to users for doing this?
Should we run this piece of code as `LoginUser` instead?

Reply:
Nice catch, if access checks are enabled the user won't have write access to the SYSTEM.TASK table, so you will need to upsert the row using User.runAsLoginUser (see MetaDataEndpointImpl.mutateRowsWithLocks).

Reply:
Will do